compiler/hash-ir/src/ir.rs

```diff
@@ -117,77 +129,183 @@ pub enum Mutability {
     Immutable,
 }
 
-/// An expression within the representation
+/// A [Statement] is a intermediate transformation step within a [BasicBlock].
 #[derive(Debug, PartialEq, Eq)]
-pub struct Expr<'i> {
-    ir_id: IrId,
-    kind: ExprKind<'i>,
+pub struct Statement<'i> {
+    /// The kind of [Statement] that it is.
+    kind: StatementKind<'i>,
+    /// The [Span] of the statement, relative to the [Body]
+    /// `source-id`.
     span: Span,
 }
 
-/// The kind of an expression
+/// Essentially a register for a value
 #[derive(Debug, PartialEq, Eq)]
-pub enum ExprKind<'i> {
+pub struct LocalDecl<'a> {
+    /// Mutability of the local.
+    mutability: Mutability,
+    /// The type of the local.
+    ty: Ty<'a>,
+}
+
+/// The addressing mode of the [Statement::AddrOf] IR statement.
+#[derive(Debug, PartialEq, Eq)]
+pub enum AddressMode {
+    /// Take the `&raw` reference of something.
+    Raw,
+    /// Take the `&` reference of something, meaning that it is reference
+    /// counted.
+    Smart,
+}
+
+/// A defined statement within the IR
+#[derive(Debug, PartialEq, Eq)]
+pub enum StatementKind<'i> {
     /// Filler kind when expressions are optimised out or removed for other
     /// reasons.
     Nop,
     /// A constant value.
     Const(Const),
+
+    /// A local variable value
+    Local(Local),
+
     /// An identity expression, essentially wrapped with an inner expression.
-    Identity(&'i Expr<'i>),
+    Identity(&'i Statement<'i>),
     /// A unary expression with a unary operator.
-    Unary(UnaryOp, &'i Expr<'i>),
+    Unary(UnaryOp, &'i Statement<'i>),
     /// A binary expression with a binary operator and two inner expressions.
-    Binary(BinOp, &'i Expr<'i>, &'i Expr<'i>),
-    /// A function call with a number of specified arguments.
-    Call(&'i Expr<'i>, &'i [Expr<'i>]),
+    Binary(BinOp, &'i Statement<'i>, &'i Statement<'i>),
     /// An index expression e.g. `x[3]`
-    Index(&'i Expr<'i>, &'i Expr<'i>),
+    Index(&'i Statement<'i>, &'i Statement<'i>),
     /// An assignment expression, a right hand-side expression is assigned to a
     /// left hand-side pattern e.g. `x = 2`
-    Assign(&'i Pat<'i>, &'i Expr<'i>),
-    /// An expression which represents a re-assignment to a pattern with a right
-    /// hand-side expression and a binary operator that combines assignment
-    /// and an operator, e.g. `x += 2`
-    AssignOp(&'i Pat<'i>, BinOp, &'i Expr<'i>),
+    Assign(Local, &'i Statement<'i>),
+
     /// An expression which is taking the address of another expression with an
     /// mutability modifier e.g. `&mut x`.
-    AddrOf(Mutability, &'i Expr<'i>),
+    Ref(Mutability, &'i Statement<'i>, AddressMode),
+
+    /// Allocate some value on the the heap using reference
+    /// counting.
+    Alloc(Local),
```

Comment:
What about non-rc? Also why does this take a local?

Reply:
what else would it take?

Reply:
What are the semantics of this? It does memcpy of the local into the location? Would it be possible to have any statement in there (prevent unnecessary copy to local first)?